contracts/extensions/contracts/src/DevUtils/OrderValidationUtils.sol

```diff
@@ -0,0 +1,173 @@
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+
+pragma solidity ^0.5.5;
+pragma experimental ABIEncoderV2;
+
+import "@0x/contracts-exchange/contracts/src/interfaces/IExchange.sol";
+import "@0x/contracts-exchange-libs/contracts/src/LibOrder.sol";
+import "@0x/contracts-utils/contracts/src/LibBytes.sol";
+import "@0x/contracts-asset-proxy/contracts/src/libs/LibAssetData.sol";
+
+
+contract OrderValidationUtils is
+    LibAssetData
+{
+    using LibBytes for bytes;
+
+    struct TraderInfo {
+        uint256 makerBalance;       // Maker's balance of makerAsset
+        uint256 makerAllowance;     // Maker's allowance to corresponding AssetProxy
+        uint256 takerBalance;       // Taker's balance of takerAsset
+        uint256 takerAllowance;     // Taker's allowance to corresponding AssetProxy
+        uint256 makerZrxBalance;    // Maker's balance of ZRX
+        uint256 makerZrxAllowance;  // Maker's allowance of ZRX to ERC20Proxy
+        uint256 takerZrxBalance;    // Taker's balance of ZRX
+        uint256 takerZrxAllowance;  // Taker's allowance of ZRX to ERC20Proxy
+    }
+
+    // solhint-disable var-name-mixedcase
+    IExchange internal EXCHANGE;
+    bytes internal ZRX_ASSET_DATA;
+    address internal ERC20_PROXY_ADDRESS;
+    // solhint-enable var-name-mixedcase
+
+    constructor (address _exchange, bytes memory _zrxAssetData)
+        public
+    {
+        EXCHANGE = IExchange(_exchange);
+        ZRX_ASSET_DATA = _zrxAssetData;
+        ERC20_PROXY_ADDRESS = EXCHANGE.getAssetProxy(ERC20_PROXY_ID);
+    }
+
+    /// @dev Fetches information for order and maker/taker of order.
+    /// @param order The order structure.
+    /// @param signature Proof that order has been created by maker.
+    /// @param takerAddress Address that will be filling the order.
+    /// @return OrderInfo, TraderInfo, and validity of signature for given order.
+    function getOrderAndTraderInfo(
+        LibOrder.Order memory order,
+        bytes memory signature,
+        address takerAddress
+    )
+        public
+        view
+        returns (
+            LibOrder.OrderInfo memory orderInfo,
+            TraderInfo memory traderInfo,
+            bool isValidSignature
+        )
+    {
+        orderInfo = EXCHANGE.getOrderInfo(order);
+        isValidSignature = EXCHANGE.isValidSignature(
+            orderInfo.orderHash,
+            order.makerAddress,
+            signature
+        );
+        traderInfo = getTraderInfo(order, takerAddress);
+        return (orderInfo, traderInfo, isValidSignature);
```

Comment:
@abandeali1 instead of returning `isValidSignature`, can't we set the `orderInfo.OrderStatus` to `SignatureInvalid`? If the signature is invalid, that should take precedence over any other status (e.g., `Fillable` doesn't make sense as an OrderStatus if the signature is invalid).

Reply:
I think the argument against this would be that `SignatureInvalid` isn't really the state of the order, but the state of the call. The order could be valid in the future if  the correct signature is provided, or may be invalid for other reasons. If you don't think that distinction is likely to be useful, we could collapse these two fields into a single `TradeState` enum.

Reply:
I think these should definitely be distinct. As is, you can pass in a valid order with an invalid signature to check the state of an order if you don't care about validating the signature. Also, an order can still be `Fillable` without a signature if the order has already been partially filled.